signer/jws.go

```diff
@@ -0,0 +1,480 @@
+package signer
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/golang-jwt/jwt/v4"
+)
+
+const (
+	JwsJsonMediaType SignatureMediaType = "application/jose+json"
+)
+
+const (
+	headerKeyExpiry      = "io.cncf.notary.expiry"
+	headerKeySigningTime = "io.cncf.notary.signingTime"
+	headerKeyCrit        = "crit"
+	headerKeyAlg         = "alg"
+	headerKeyCty         = "cty"
+)
+
+// jwsInternalEnvelope represents the jwsEnvelope-JSON envelope
+type jwsEnvelope struct {
+	internalEnv *jwsInternalEnvelope
+}
+
+func newJWSEnvelopeFromBytes(envelopeBytes []byte) (*jwsEnvelope, error) {
+	jwsInternal, err := newJwsInternalEnvelopeFromBytes(envelopeBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return &jwsEnvelope{internalEnv: jwsInternal}, nil
+}
+
+func newJWSEnvelope() *jwsEnvelope {
+	return &jwsEnvelope{}
+}
+
+func (jws *jwsEnvelope) validateIntegrity() error {
+	if jws.internalEnv == nil {
+		return SignatureNotFoundError{}
+	}
+
+	if len(jws.internalEnv.Header.CertChain) == 0 {
+		return MalformedSignatureError{msg:"malformed leaf certificate"}
+	}
+
+	cert, err := x509.ParseCertificate(jws.internalEnv.Header.CertChain[0])
+	if err != nil {
+		return MalformedSignatureError{msg:"malformed leaf certificate"}
+	}
+
+	// verify JWT
+	compact := strings.Join([]string{jws.internalEnv.Protected, jws.internalEnv.Payload, jws.internalEnv.Signature}, ".")
+	return verifyJWT(compact, cert.PublicKey)
+}
+
+func (jws *jwsEnvelope) signPayload(req SignRequest) ([]byte, error) {
+	leafPublicKey := req.CertificateChain[0].PublicKey
+	var m map[string]interface{}
+	if err := json.Unmarshal(req.Payload, &m); err != nil {
+		return nil, err
+	}
+	signingMethod, err := getSigningMethod(leafPublicKey)
+	if err != nil {
+		return nil, err
+	}
+
+	signedAttrs := getSignedAttrs(req, signingMethod)
+	compact, err := signJWT(m, signedAttrs, signingMethod, req.SignatureProvider)
+	if err != nil {
+		return nil, err
+	}
+
+	j, err := generateJws(compact, req)
+	if err != nil {
+		return nil, err
+	}
+
+	b, err := json.Marshal(j)
+	if err != nil {
+		return nil, err
+	}
+	jws.internalEnv = j
+	return b, nil
+}
+
+func (jws *jwsEnvelope) getSignerInfo() (*SignerInfo, error) {
+	signInfo := SignerInfo{}
+	if jws.internalEnv == nil {
+		return nil, SignatureNotFoundError{}
+	}
+
+	// parse payload
+	payload, err := base64.RawURLEncoding.DecodeString(jws.internalEnv.Payload)
+	if err != nil {
+		return nil, err
+	}
+	signInfo.Payload = payload
+
+	// parse protected headers
+	protected, err := base64.RawURLEncoding.DecodeString(jws.internalEnv.Protected)
+	if err != nil {
+		return nil, err
+	}
+	var pHeaders map[string]interface{}
+	if err = json.Unmarshal(protected, &pHeaders); err != nil {
+		return nil, err
+	}
+	if err := populateProtectedHeaders(pHeaders, &signInfo); err != nil {
+		return nil, err
+	}
```

Comment:
I'm having troubles following the logic in `populateProtectedHeaders`. Why don't we decode the protected header using a `struct` with the appropriate properties and json tags instead of unmarshaling into a map and then manually populating everything? It would be less complex and easier to maintain.

Reply:
Notary v2 allows users to add custom signed attributes and in the case of JWS these custom signed attributes will go into protected headers.
Since we don't know the keys for these custom attributes we cannot use `struct` and that's the reason we need to unmarshall protected headers to a map and then manually populate everything

Reply:
Good point. If that's the case, I would prefer to unmarshall into a typed struct for the known fields and then unmarshall again into a map deleting the known fields.

This way we can get rid of `populateProtectedHeaders`, aka less code to maintain, and we only pay an, and probably unnoticeable, performance penalty.

Reply:
Updated code to unmarshall into a typed struct for the known fields and then unmarshall again into a map deleting the known fields.